vm/systemSmartContracts/staking.go

```diff
@@ -88,16 +113,206 @@ func (r *stakingSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCod
+	if startRound > endRound {
+		return stake
+	}
+	if endRound-startRound < r.numRoundsWithoutBleed {
+		return stake
+	}
+
+	totalRoundsToBleed := endRound - startRound - r.numRoundsWithoutBleed
+	totalPercentageToBleed := float64(totalRoundsToBleed) * r.bleedPercentagePerRound
+	totalPercentageToBleed = math.Max(totalPercentageToBleed, r.maximumPercentageToBleed)
+
+	bleedValue := big.NewInt(0).Neg(getPercentageOfValue(stake, totalPercentageToBleed))
+	stakeAfterBleed := big.NewInt(0).Add(stake, bleedValue)
+
+	if stakeAfterBleed.Cmp(big.NewInt(0)) < 0 {
+		stakeAfterBleed = big.NewInt(0)
+	}
+
+	return stakeAfterBleed
+}
+
+func (r *stakingSC) changeValidatorKey(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.stakeAccessAddr) {
+		log.Debug("stake function not allowed to be called by", "address", args.CallerAddr)
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		return vmcommon.UserError
+	}
+
+	oldKey := args.Arguments[0]
+	newKey := args.Arguments[1]
+	if len(oldKey) != len(newKey) {
+		return vmcommon.UserError
+	}
+
+	stakedData, err := r.getOrCreateRegisteredData(oldKey)
+	if err != nil {
+		return vmcommon.UserError
+	}
+	if len(stakedData.RewardAddress) == 0 {
+		// if not registered this is not an error
+		return vmcommon.Ok
+	}
+
+	r.eei.SetStorage(oldKey, nil)
+	err = r.saveStakingData(newKey, stakedData)
+	if err != nil {
+		return vmcommon.UserError
+	}
+
+	return vmcommon.Ok
+}
+
+func (r *stakingSC) changeRewardAddress(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.stakeAccessAddr) {
+		log.Debug("stake function not allowed to be called by", "address", args.CallerAddr)
+		return vmcommon.UserError
+	}
+	if len(args.Arguments) < 2 {
+		return vmcommon.UserError
+	}
+
+	newRewardAddress := args.Arguments[0]
+	if len(newRewardAddress) != len(args.CallerAddr) {
+		return vmcommon.UserError
+	}
+
+	for i := 1; i < len(args.Arguments); i++ {
+		blsKey := args.Arguments[i]
+		stakedData, err := r.getOrCreateRegisteredData(blsKey)
+		if err != nil {
+			return vmcommon.UserError
+		}
+		if len(stakedData.RewardAddress) == 0 {
+			continue
+		}
+
+		stakedData.RewardAddress = newRewardAddress
+		err = r.saveStakingData(blsKey, stakedData)
+		if err != nil {
+			return vmcommon.UserError
+		}
+	}
+
+	return vmcommon.Ok
+}
+
 func (r *stakingSC) unJail(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if !bytes.Equal(args.CallerAddr, r.stakeAccessAddr) {
+		log.Debug("stake function not allowed to be called by", "address", args.CallerAddr)
+		return vmcommon.UserError
+	}
+
+	for _, argument := range args.Arguments {
+		stakedData, err := r.getOrCreateRegisteredData(argument)
+		if err != nil {
+			return vmcommon.UserError
+		}
+		if len(stakedData.RewardAddress) == 0 {
+			return vmcommon.UserError
+		}
+
+		stakedData.StakeValue = r.calculateStakeAfterBleed(
```

Comment:
What happens if the remaining staked value is below the threshold minimum stake value?

Reply:
will not be selected in auction.